File: src/types/workspace-health.ts
import type { WorkspacePort } from './workspace-ports';

export interface WorkspaceConflict {
  workspaceIdA: string;
  workspaceIdB: string;
  sharedFiles: string[];
  fileCount: number;
}

export interface WorkspaceConflicts {
  conflicts: WorkspaceConflict[];
  conflictingWorkspaceIds: string[];
}

export interface WorkspaceTerminalHealth {
  sessionId: string;
  title: string;
  kind: string;
  profile: string;
  status: string;
  backend: string;
  attached: boolean;
  stale: boolean;
  recoveryStatus: 'recoverable_running' | 'orphaned' | 'stale_silent' | 'interrupted' | 'closed' | string;
  lastOutputAt?: string | null;
  recommendedAction: string;
  /** Unix timestamp (seconds) when silence crossed the stuck threshold. Only set for running agent sessions. */
  stuckSince?: string | null;
}

export interface WorkspaceHealth {
  workspaceId: string;
  status: 'healthy' | 'needs_attention' | 'idle' | string;
  terminals: WorkspaceTerminalHealth[];
  ports: WorkspacePort[];
  warnings: string[];
}

export interface WorkspaceSessionRecoveryResult {
  workspaceId: string;
  closedSessions: number;
  skippedSessions: number;
  actions: WorkspaceSessionRecoveryAction[];
  warnings: string[];
}

export interface WorkspaceSessionRecoveryAction {
  sessionId: string;
  title: string;
  action: 'closed' | 'skipped' | 'failed' | 'resumed' | 'marked_interrupted' | string;
  reason: string;
}

export interface ApplyWorkspaceSessionRecoveryInput {
  workspaceId: string;
  sessionId: string;
  action: 'resume_tracking' | 'mark_interrupted' | 'close_session' | string;
  reason?: string | null;
}

File: src-tauri/src/models/workspace_health.rs
use serde::{Deserialize, Serialize};

use crate::models::WorkspacePort;

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct WorkspaceTerminalHealth {
    pub session_id: String,
    pub title: String,
    pub kind: String,
    pub profile: String,
    pub status: String,
    pub backend: String,
    pub attached: bool,
    pub stale: bool,
    pub last_output_at: Option<String>,
    pub recommended_action: String,
    /// Unix timestamp (seconds) when the agent was first detected as having no output.
    /// Only set for running agent sessions silent for > STUCK_THRESHOLD_SECS.
    pub stuck_since: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct WorkspaceHealth {
    pub workspace_id: String,
    pub status: String,
    pub terminals: Vec<WorkspaceTerminalHealth>,
    pub ports: Vec<WorkspacePort>,
    pub warnings: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct WorkspaceSessionRecoveryResult {
    pub workspace_id: String,
    pub closed_sessions: u32,
    pub skipped_sessions: u32,
    pub warnings: Vec<String>,
}
